File: 0040-combination-sum-ii/0040-combination-sum-ii.go
//import "sort"

func combinationSum2(candidates []int, target int) [][]int {
    sort.Ints(candidates)
    var result [][]int
    backtrack(candidates, target, 0, []int{}, &result)
    return result
}

func backtrack(candidates []int, target int, start int, path []int, result *[][]int) {
    if target == 0 {
        // Make a deep copy of the result and append to the list.
        tmp := make([]int, len(path))
        copy(tmp, path)
        *result = append(*result, tmp)
        return
    }
    
    for i := start; i < len(candidates); i++ {
        // If the current candidate is greater than the target, break the loop.
        if candidates[i] > target {
            break
        }
        // If the current candidate is the same as the previous one, skip it.
        if i > start && candidates[i] == candidates[i - 1] {
            continue
        }
        // Include the candidate in the path.
        path = append(path, candidates[i])
        // Recurse, reducing the target by the chosen candidate's value.
        backtrack(candidates, target - candidates[i], i + 1, path, result)
        // Exclude the candidate from the path.
        path = path[:len(path) - 1]
    }
}

File: 0040-combination-sum-ii/0040-combination-sum-ii.java
class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> res = new ArrayList<>();
        findCombinations(candidates, target, 0, new ArrayList<>(), res);
        return res;
    }
    
    private void findCombinations(int[] candidates, int target, int index, List<Integer> current, List<List<Integer>> res) {
        if(target == 0) {
            res.add(new ArrayList<>(current));
            return;
        }
        if(target < 0) return;
        for(int i = index; i < candidates.length; i++) {
            if(i == index || candidates[i] != candidates[i - 1]) {
                current.add(candidates[i]);
                findCombinations(candidates, target - candidates[i], i + 1, current, res);
                current.remove(current.size() - 1);
            }
        }
    }
}
